File: 0189-rotate-array/0189-rotate-array.cpp
#include <bits/stdc++.h>

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        
        k %= nums.size();
        
        int n = nums.size();
        
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
        
    
        
        
    }
    
    
    
    void reverse(vector<int>& nums, int left, int right) {
        while (left < right) {
            swap(nums[left], nums[right]);
            left++;
            right--;
        }
    }
};

File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k %= n
        
        # reverse list
        nums[:len(nums)] = nums[:len(nums)][::-1]
        
        # reverse 2 parts
        nums[:k] = nums[:k][::-1]
        nums[k:] = nums[k:][::-1]
        